src/Compilers/CSharp/Test/Semantic/Semantics/InterceptorsTests.cs

```diff
@@ -0,0 +1,1357 @@
+    // * Intercept with extern method
+    // * Intercept with abstract or interface method
+    // * Intercept an abstract or interface method
+    // * Intercept a virtual or overridden method
+    // * Intercept a non-extension call to a static method with a static method when one or both are extension methods
+    // * Intercept a struct instance method with an extension method with by-value / by-ref this parameter
+    // * An explicit interface implementation marked as interceptable
+    // * Intercept a generic method call when the type parameters are / are not substituted
+
+    // PROTOTYPE(ic): test a valid case with large numbers of interceptors. (EndToEndTests?)
+
+    [Fact]
+    public void InterceptableInstanceMethod_InterceptorExtensionMethod()
+    {
+        var source = """
+            using System.Runtime.CompilerServices;
+            using System;
+
+            class C
+            {
+                [Interceptable]
+                public C InterceptableMethod(string param) { Console.Write("interceptable " + param); return this; }
+            }
+
+            static class Program
+            {
+                public static void Main()
+                {
+                    var c = new C();
+                    c.InterceptableMethod("call site");
+                }
+            }
+
+            static class D
+            {
+                [InterceptsLocation("Program.cs", 15, 11)]
+                public static C Interceptor1(this C i1, string param) { Console.Write("interceptor " + param); return i1; }
+            }
+            """;
+        var verifier = CompileAndVerify(new[] { (source, "Program.cs"), s_attributesSource }, expectedOutput: "interceptor call site");
+        verifier.VerifyDiagnostics();
+    }
+
+    [Fact]
+    public void QualifiedNameAtCallSite()
+    {
+        var source = """
+            using System.Runtime.CompilerServices;
+            using System;
+
+            class C
+            {
+                [Interceptable]
+                public static C InterceptableMethod(C c, string param) { Console.Write("interceptable " + param); return c; }
+            }
+
+            static class Program
+            {
+                public static void Main()
+                {
+                    var c = new C();
+                    C.InterceptableMethod(c, "call site");
+                }
+            }
+
+            static class D
+            {
+                [InterceptsLocation("Program.cs", 15, 11)]
+                public static C Interceptor1(C c, string param) { Console.Write("interceptor " + param); return c; }
+            }
+            """;
+        var verifier = CompileAndVerify(new[] { (source, "Program.cs"), s_attributesSource }, expectedOutput: "interceptor call site");
+        verifier.VerifyDiagnostics();
+    }
+
+    [Fact]
+    public void InterceptableStaticMethod_InterceptorExtensionMethod()
```

Comment:
Spec: What is the benefit of allowing this scenario? I'd be tempted to go with stricter rules: static-ness must match, extension-ness must match.

Reply:
I did this because an extension method can be called in either normal form or reduced form. However, I don't feel strongly here.

Also wanted to note that the bound tree for extension calls is quite similar whether they are called in normal or reduced form--that is, I don't think disallowing this difference reduces implementation complexity.